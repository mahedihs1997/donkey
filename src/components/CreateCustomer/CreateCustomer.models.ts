export interface CustomerValues {
  name: string;
  contact: string;
  lineOne: string;
  lineTwo: string;
  city: string;
  country: string;
  postcode: string;
  mobilePhone: string;
  workPhone: string;
  accountType: string;
}

export interface CreateCustomerProps {
  onSuccess: () => void;
}

export interface CreateCustomerRefs {
  handleClose: () => void;
  handleOpen: () => void;
}
