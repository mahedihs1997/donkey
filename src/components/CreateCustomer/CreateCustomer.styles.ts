import { makeStyles } from '@material-ui/core/styles';

export const useStyles = makeStyles({
  root: {
    display: 'flex',
  },
  form: {
    '&:first-child': {
      marginRight: 30,
    },
  },
  button: {
    marginTop: 30,
  },
});
