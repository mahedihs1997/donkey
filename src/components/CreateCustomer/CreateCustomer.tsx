import FormikSelect from '@components/FormikSelect';
import FormikTextInput from '@components/FormikTextInput';
import Modal from '@components/Modal';
import { Button } from '@material-ui/core';
import { Form, Formik } from 'formik';
import React from 'react';
import { toast } from 'react-toastify';

import {
  CreateCustomerProps,
  CreateCustomerRefs,
} from './CreateCustomer.models';
import { useStyles } from './CreateCustomer.styles';
import { validateCreateCustomer } from './CreateCustomer.validate';

const CreateCustomer = React.forwardRef<
  CreateCustomerRefs,
  CreateCustomerProps
>((props, ref) => {
  const { onSuccess } = props;
  const [open, setOpen] = React.useState(false);

  const handleClose = () => {
    setOpen(false);
  };
  const handleOpen = () => {
    setOpen(true);
  };

  React.useImperativeHandle(ref, () => ({
    handleClose,
    handleOpen,
  }));

  const classes = useStyles();

  const notifyError = () => {
    toast.error('Creating customer failed ðŸ˜‹');
  };

  const notifySuccess = () => {
    toast.success('Creating customer succeeded ðŸ˜‹');
  };

  return (
    <Modal open={open} onClose={handleClose} title="Create User">
      <Formik
        initialValues={{
          name: '',
          contact: '',
          lineOne: '',
          lineTwo: '',
          city: '',
          country: '',
          postcode: '',
          mobilePhone: '',
          workPhone: '',
          accountType: 'Casual',
        }}
        validate={validateCreateCustomer}
        onSubmit={async (values, actions) => {
          try {
            console.log('Values', values);
            notifySuccess();
            onSuccess();
            actions.resetForm();
          } catch (e) {
            notifyError();
            console.error('Creating user failed with error: ', e);
          }
        }}
      >
        {(props) => (
          <Form>
            <div className={classes.root}>
              <div className={classes.form}>
                <FormikTextInput name="name" label="Name" />
                <FormikTextInput
                  name="contact"
                  label="Contact Name (Optional)"
                />
                <FormikTextInput name="mobilePhone" label="Mobile Phone" />
                <FormikTextInput
                  name="workPhone"
                  label="Work Phone (Optional)"
                />
                <FormikSelect
                  name="accountType"
                  label="Account Type"
                  placeholder="Please choose a position"
                  data={[
                    { value: 'Casual', label: 'Casual' },
                    { value: 'Account Holder', label: 'Account Holder' },
                  ]}
                />
              </div>
              <div className={classes.form}>
                <FormikTextInput name="lineOne" label="Line One" />
                <FormikTextInput name="lineTwo" label="Line Two (Optional)" />
                <FormikTextInput name="city" label="City" />
                <FormikTextInput name="country" label="Country" />
                <FormikTextInput
                  name="postcode"
                  label="Postcode (If applicable)"
                />
              </div>
            </div>
            <Button
              variant="contained"
              color="primary"
              type="submit"
              disabled={props.isSubmitting}
              className={classes.button}
              fullWidth={true}
            >
              Create New Customer
            </Button>
          </Form>
        )}
      </Formik>
    </Modal>
  );
});

CreateCustomer.displayName = 'Create Customer';

export default CreateCustomer;
