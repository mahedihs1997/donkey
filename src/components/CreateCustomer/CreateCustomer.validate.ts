import { FormikErrors } from 'formik';

import { CustomerValues } from './CreateCustomer.models';

export const validateCreateCustomer = (values: CustomerValues) => {
  const errors: FormikErrors<CustomerValues> = {};
  if (!values.name) {
    errors.name = 'Please enter a name';
  }

  if (!values.lineOne) {
    errors.lineOne = 'Please enter a street';
  }

  if (!values.city) {
    errors.city = 'Please enter a city';
  }

  if (!values.country) {
    errors.country = 'Please enter a country';
  }

  if (!values.mobilePhone) {
    errors.mobilePhone = 'Please enter a phone number';
  }

  return errors;
};
