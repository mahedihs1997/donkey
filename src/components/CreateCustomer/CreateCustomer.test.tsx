import React from 'react';
import TestRenderer from 'react-test-renderer';

import CreateCustomer from '.';

describe('<CreateCustomer />', () => {
  it('should render without throwing an error and match snapshot', () => {
    const testRenderer = TestRenderer.create(
      <CreateCustomer onSuccess={jest.fn()} />
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
});
