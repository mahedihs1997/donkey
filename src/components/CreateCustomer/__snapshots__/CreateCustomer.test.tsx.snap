// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateCustomer /> should render without throwing an error and match snapshot 1`] = `null`;
